feat(http): add checker for required interceptor registration

Sepandard opened the pull request:
resolves #57409

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] Minimal code changes to integrate new functionality or tests
- [x] Code maintains backward compatibility
- [x] All tests pass successfully

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A

## What is the new behavior?
<!-- Please describe the new behavior after the changes. -->
- Added utility functions and corresponding test cases for setting XSRF tokens and tagging HTTP requests.
- Improved test coverage with minimal code changes and ensured backward compatibility.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information
<!-- Any additional information that could be useful for reviewers. -->
- All tests are designed to integrate seamlessly without affecting existing functionality.

JeanMeche commented on packages/common/http/src/provider.ts:
```diff
@@ -174,6 +190,19 @@ const LEGACY_INTERCEPTOR_FN = new InjectionToken<HttpInterceptorFn>(
+export function ensureInterceptorRegistered(interceptorName: string): void {
+  const checkedInterceptors = inject(CHECKED_INTERCEPTORS);
+  if (!checkedInterceptors.has(interceptorName)) {
+    throw new Error(
```
Ideally we should make this a `RuntimeError`, so we can remove the message string on prod builds. (See other usages of `RuntimeError`).

JeanMeche commented on packages/common/http/src/provider.ts:
```diff
@@ -161,6 +174,9 @@ export function withInterceptors(
   return makeHttpFeature(
     HttpFeatureKind.Interceptors,
     interceptorFns.map((interceptorFn) => {
+      const interceptorName = interceptorFn.name || 'AnonymousInterceptor';
```
AFAIK, it isn't safe to reach-out for `name`. 
We probably should/can store the interceptor definition itself.

Sepandard replied:
I changed it, It's checking the interceptor definition itself now.

Sepandard committed:
feat(http): add checker for required interceptor registration

Introduced a CHECKED_INTERCEPTORS token to track registered interceptors.
Added logic in withInterceptors to register interceptor names and validate their presence.
Added ensureInterceptorRegistered utility to enforce the registration of mandatory interceptors.
update tests and document

resolves #57409

thePunderWoman reviewed (approved):
LGTM! Thank you!

reviewed-for: fw-general, public-api, angular-dev

kirjs reviewed (approved):
Reviewed-for: public-api
